docs/network-binding-plugin.md

```diff
@@ -0,0 +1,342 @@
+to the domain.
+
+The currently supported domain attachment type is `tap` (v1.1.1) which
+builds a domain interface configuration that points to the tap/macvtap
+existing interface.
+
+Such a binding plugin assumes that the CNI used for the network connectivity
+exposes in the pod a `tap` or `macvtap` (type) interface with a name corresponding
+to the hashed network name:
+
+- `pod<hash network name>` (or plain `eth0` for the primary network)
+- `tap<hash network name>` (or plain `tap0` for the primary network)
+
+Kubevirt virt-controller will define these names on the pod multus annotation,
+therefore there is no special action needed from the plugin author except
+choosing a CNI that confirms with the CNI standard spec.
+
+For clarification, the CNI used for the connectivity is the one defined
+in the `NetworkAttachmentDefinition` object referenced from the VM network spec.
+
+### Configuration example
+
+Given the following registered binding plugin:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  name: kubevirt
+  namespace: kubevirt
+spec:
+  configuration:
+    network:
+      Binding:
+        my-zero-code-binding:
+          domainAttachment: tap
+```
+
+The VM interface/network spec will include a ref to the binding plugin
+and the `NetworkAttachmentDefinition` named `mynad`:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: myvm
+spec:
+  template:
+    spec:
+      domain:
+        devices:
+          interfaces:
+            - name: mynetwork
+              binding:
+                name: my-zero-code-binding
+      networks:
+      - name: mynetwork
+        multus:
+          networkName: mmynad
+```
+
+The `mynad` NetworkAttachmentDefinition defines `cni-name` as the CNI plugin
+binary:
+
+```yaml
+kind: NetworkAttachmentDefinition
+apiVersion: k8s.cni.cncf.io/v1
+metadata:
+  name: mynad
+spec:
+  config: '{
+      "cniVersion": "0.3.1",
+      "name": "some-name",
+      "type": "cni-name",
+    }'
+```
+
+Kubevirt will eventually create a domain interface configuration that
+points to the pod interface (e.g. `pod12345678`).
+This is a snippet of the configuration:
+
+```xml
+<interface type='ethernet'>
+   <alias name='ua-mynetwork'/>
+   <target dev='pod12345678' managed='no'/>
+   <model type='virtio-non-transitional'/>
+   <mac address='12:34:56:78:9a:bc'/>
+   <mtu size='1500'/>
+   <rom enabled='no'/>
+</interface>
+```
+
+A classic example of such a plugin is thehttps://github.com/kubevirt/kubevirt/blob/main/pkg/hooks/hooks.go#L31
+[macvtap](https://kubevirt.io/user-guide/virtual_machines/net_binding_plugins/macvtap/)
+plugin.
+
+## The Sidecar Plugin
+
+When a standard domain attachment requires customization,
+or when additional services are needed (e.g. DHCP), a sidecar container
+may be executed to integrate with Kubevirt.
+
+The sidecar container runs in parallel to the virt-launcher container
+which runs the hypervisor (libvirt and qemu).
+
+### Sidecar Protocol
```

Comment:
I think the content that related to Kubevirt sidecar hooks should be moved to dedicated file for that topic.
They we could reference it in this doc where needed.
I know there is no Kubevirt sidecar hook doc but what you wrote so far cover most of it.

Reply:
Perhaps, but it is not in my scope at the moment.
AFAIK the sidecar hook examples use a shim layer which in our context we do not use.

I have also not covered all APIs of the protocol, only the ones we found useful for our need.
By extracting this on its own, it will require a full coverage and explanations here that we do not need all.

Bottom line, for now I prefer to be focused only on the network binding plugin needs.